gazebo-tooling/release-tools
# Add new job to run outdated gazebo jenkins jobs

### Description

This PR adds a new job that will automate the manual process of running outdated jobs.

The automation this new job adds will clear all human error to maintain the gazebo buildfarm jobs up to date by running every night and checking all the jobs buildfarmers are tracking.

After merging this PR, there will be a new job called `_outdated_job_runner`. It will run nightly (as defined in [`CRON_NIGHTLY_NODES`](https://github.com/gazebo-tooling/release-tools/blob/master/jenkins-scripts/dsl/_configs_/Globals.groovy#L18)), and it will trigger a maximun of 7 jobs per operating system to prevent buildfarm overload if more jobs are outdated (this was internally defined by the buildfarmer team).

<details>

<summary>
Some extra context:
</summary>

Currently, Gazebo jobs only run with `scm` triggers. Sometimes, the repositories don't change in a lot of time, and we don't know if they're still working as they should (the main concern are dependencies. A broken python package can lead to many problems if not noticed early). Right now, we have an 8 days threshold for a job to be considered "outdated". When this happens, buildfarmer daily report (e.g., [2023-06-21.md](https://github.com/osrf/buildfarmer/blob/main/2023/25/2023-06-21.md)) will show a message saying `JOB HAS NOT BEEN UPDATED RECENTLY`. Then, we check and rebuild all outdated jobs by hand.

</details>

### Testing

The generated job was named [`_nightly_job_runner`](https://build.osrfoundation.org/job/_nightly_job_runner/) before. It had some test configurations that I used to verify that it was running correctly. For example: 
* Build to check what were the outdated jobs ([_nightly_job_runner#17](https://build.osrfoundation.org/job/_nightly_job_runner/17/console))
* Triggered build by _nightly_job_runner: [ign_gui-ign-6-win#75](https://build.osrfoundation.org/job/ign_gui-ign-6-win/75/) (you can see the upstream project that built it and all the rebuilds before it

Now I'm using [`_outdated_job_runner`](https://build.osrfoundation.org/job/_outdated_job_runner/) to test the new configuration:
* Build that triggered a `_test_job_from_dsl` build ([_outdated_job_runner#4](https://build.osrfoundation.org/job/_outdated_job_runner/4/))
  <details>
  
  I used the following script to trigger a `_test_job_from_dsl` build:

  ```
  def job = "_test_job_from_dsl";

  println('Running: ' + job)
  runJob(job)
  ```

  </details>
* Current dry-run output (using modification of current script to not trigger all the builds): ([_outdated_job_runnger#5](https://build.osrfoundation.org/job/_outdated_job_runner/5/console))
  <details>
  
  <summary>
  Log output explained:
  </summary>
  <br/>

   <details>
  <summary>
  All linux outdated jobs (7 in total): 
  </summary>

  ```
  Linux outdated jobs: [ignition_cmake-ci-ign-cmake2-focal-amd64, ignition_launch-ci-gz-launch6-focal-amd64, ignition_msgs-ci-ign-msgs8-focal-amd64, ignition_plugin-ci-gz-plugin2-focal-amd64, ignition_rendering-ci-ign-rendering6-focal-amd64, ignition_utils-ci-gz-utils2-focal-amd64, sdformat-ci-sdformat9-focal-amd64]
  ```

  </details>

  <details>
  <summary>
  All homebrew outdated jobs (10 in total):
  </summary>

  ```
  Homebrew outdated jobs: [ignition_common-ci-ign-common3-homebrew-amd64, ignition_gui-ci-ign-gui3-homebrew-amd64, ignition_gui-ci-ign-gui6-homebrew-amd64, ignition_msgs-ci-ign-msgs8-homebrew-amd64, ignition_plugin-ci-gz-plugin2-homebrew-amd64, ignition_rendering-ci-ign-rendering6-homebrew-amd64, ignition_tools-ci-gz-tools2-homebrew-amd64, ignition_utils-ci-gz-utils2-homebrew-amd64, sdformat-ci-sdformat9-homebrew-amd64]
  ```

  </details>

  <details>
  <summary>
  All windows outdated jobs (4 in total):
  </summary>

  ```
  Windows outdated jobs: [ign_msgs-ign-8-win, ign_rendering-ign-6-win, ign_utils-gz-2-win, sdformat-sdf-12-win]
  ```

  </details>

  <details>
  <summary>
  Final list of jobs to run (7 from linux, 7 from homebrew, 4 from windows, 18 in total):
  </summary>

  ```
  Jobs to run: [ignition_cmake-ci-ign-cmake2-focal-amd64, ignition_launch-ci-gz-launch6-focal-amd64, ignition_msgs-ci-ign-msgs8-focal-amd64, ignition_plugin-ci-gz-plugin2-focal-amd64, ignition_rendering-ci-ign-rendering6-focal-amd64, ignition_utils-ci-gz-utils2-focal-amd64, sdformat-ci-sdformat9-focal-amd64, ignition_common-ci-ign-common3-homebrew-amd64, ignition_gui-ci-ign-gui3-homebrew-amd64, ignition_gui-ci-ign-gui6-homebrew-amd64, ignition_msgs-ci-ign-msgs8-homebrew-amd64, ignition_plugin-ci-gz-plugin2-homebrew-amd64, ignition_rendering-ci-ign-rendering6-homebrew-amd64, ignition_tools-ci-gz-tools2-homebrew-amd64, ign_msgs-ign-8-win, ign_rendering-ign-6-win, ign_utils-gz-2-win, sdformat-sdf-12-win]
  ```

  </details>

  <details>
  <summary>
  Dry run output:
  </summary>

  ```
  Running: ignition_cmake-ci-ign-cmake2-focal-amd64
  Running: ignition_launch-ci-gz-launch6-focal-amd64
  Running: ignition_msgs-ci-ign-msgs8-focal-amd64
  Running: ignition_plugin-ci-gz-plugin2-focal-amd64
  Running: ignition_rendering-ci-ign-rendering6-focal-amd64
  Running: ignition_utils-ci-gz-utils2-focal-amd64
  Running: sdformat-ci-sdformat9-focal-amd64
  Running: ignition_common-ci-ign-common3-homebrew-amd64
  Running: ignition_gui-ci-ign-gui3-homebrew-amd64
  Running: ignition_gui-ci-ign-gui6-homebrew-amd64
  Running: ignition_msgs-ci-ign-msgs8-homebrew-amd64
  Running: ignition_plugin-ci-gz-plugin2-homebrew-amd64
  Running: ignition_rendering-ci-ign-rendering6-homebrew-amd64
  Running: ignition_tools-ci-gz-tools2-homebrew-amd64
  Running: ign_msgs-ign-8-win
  Running: ign_rendering-ign-6-win
  Running: ign_utils-gz-2-win
  ```
  
  </details>

  </details>
### Points of discussion

1. There are a lot of deprecated jobs that haven't run since 2021 and 2022. To filter these, I added a list of tracked jobs to the code: https://github.com/gazebo-tooling/release-tools/blob/ca021711dc56f29ed064acf82b8c42aebe9221b3/jenkins-scripts/tools/outdated-job-runner.groovy#L5-L8 (FYI: this list contains all the names of jobs that are in [Ignition.md](https://github.com/osrf/buildfarmer/blob/main/Ignition.md) dashboard that are not debbuilders)
    It would be great if I can modify it to be dynamically defined (e.g., all jobs that match to a set of predefined regex) 
2.  I couldn't find an easy way to run a job with its default parameters, so the workaround is: get the default parameters and pass them as new parameters. https://github.com/gazebo-tooling/release-tools/blob/ca021711dc56f29ed064acf82b8c42aebe9221b3/jenkins-scripts/tools/outdated-job-runner.groovy#L13-L19
